File: Masjidly - Official Masjid Prayer Times/Features/Settings/SettingsViewModel.swift
import Foundation
import Observation
import UserNotifications

enum TestNotificationType: String, CaseIterable {
    case adhan = "Adhan"
    case iqamah = "Iqamah"
    case reminder = "Reminder"
    case all = "All Three"
}

@Observable
@MainActor
final class SettingsViewModel {
    private let repository: any PrayerRepository
    private let settings: SettingsStore
    private let notificationScheduler: any PrayerNotificationScheduling
    private let diskCache: PrayerTimesDiskCache

    var mosques: [Mosque] = []
    var loadError: String?

    init(
        repository: any PrayerRepository,
        settings: SettingsStore,
        notificationScheduler: any PrayerNotificationScheduling,
        diskCache: PrayerTimesDiskCache
    ) {
        self.repository = repository
        self.settings = settings
        self.notificationScheduler = notificationScheduler
        self.diskCache = diskCache
    }

    func load() async {
        loadError = nil

        // Hydrate from cache first for instant display.
        if let cached = diskCache.loadMosques() {
            mosques = MosqueDefaults.visibleMosques(cached)
        }

        // Then fetch from Convex and save to cache.
        do {
            let all = try await repository.listMosques()
            mosques = MosqueDefaults.visibleMosques(all)
            try? diskCache.saveMosques(all)
        } catch {
            loadError = error.localizedDescription
        }
    }

    func selectMosque(_ mosque: Mosque) async {
        settings.selectedMosqueId = mosque.id
        settings.selectedMosqueSlug = mosque.slug
        await applyNotificationPolicy()
    }

    func onNotificationsChanged() async {
        await applyNotificationPolicy()
    }

    func fireTestNotification(_ type: TestNotificationType) async {
        do {
            _ = try await notificationScheduler.requestAuthorizationIfNeeded()
        } catch {
            return
        }
        let auth = await UNUserNotificationCenter.current().notificationSettings().authorizationStatus
        guard auth == .authorized || auth == .provisional else { return }

        let slug = settings.selectedMosqueSlug ?? ""

        switch type {
        case .adhan:
            let copy = PrayerNotificationContent.adhanCopy(prayerKey: "maghrib", isFriday: false)
            await enqueueTestNotification(
                title: copy.title,
                body: copy.body,
                categoryIdentifier: PrayerNotificationContent.CategoryID.adhan,
                sound: PrayerNotificationContent.sound(for: settings.notifications, channel: .adhan),
                userInfo: PrayerNotificationContent.debugUserInfo(kind: .adhan, prayerKey: "maghrib", mosqueSlug: slug)
            )
        case .iqamah:
            let copy = PrayerNotificationContent.iqamahCopy(prayerKey: "maghrib", isFriday: false)
            await enqueueTestNotification(
                title: copy.title,
                body: copy.body,
                categoryIdentifier: PrayerNotificationContent.CategoryID.iqamah,
                sound: PrayerNotificationContent.sound(for: settings.notifications, channel: .iqamah),
                userInfo: PrayerNotificationContent.debugUserInfo(kind: .iqamah, prayerKey: "maghrib", mosqueSlug: slug)
            )
        case .reminder:
            let copy = PrayerNotificationContent.beforeAdhanReminderCopy(prayerKey: "maghrib", isFriday: false, minutes: 10)
            await enqueueTestNotification(
                title: copy.title,
                body: copy.body,
                categoryIdentifier: PrayerNotificationContent.CategoryID.reminder,
                sound: PrayerNotificationContent.sound(for: settings.notifications, channel: .reminder),
                userInfo: PrayerNotificationContent.debugUserInfo(kind: .reminderBeforeAdhan, prayerKey: "maghrib", mosqueSlug: slug)
            )
        case .all:
            let adhan = PrayerNotificationContent.adhanCopy(prayerKey: "maghrib", isFriday: false)
            await enqueueTestNotification(
                title: adhan.title,
                body: adhan.body,
                categoryIdentifier: PrayerNotificationContent.CategoryID.adhan,
                sound: PrayerNotificationContent.sound(for: settings.notifications, channel: .adhan),
                userInfo: PrayerNotificationContent.debugUserInfo(kind: .adhan, prayerKey: "maghrib", mosqueSlug: slug)
            )
            let iq = PrayerNotificationContent.iqamahCopy(prayerKey: "maghrib", isFriday: false)
            await enqueueTestNotification(
                title: iq.title,
                body: iq.body,
                categoryIdentifier: PrayerNotificationContent.CategoryID.iqamah,
                sound: PrayerNotificationContent.sound(for: settings.notifications, channel: .iqamah),
                userInfo: PrayerNotificationContent.debugUserInfo(kind: .iqamah, prayerKey: "maghrib", mosqueSlug: slug)
            )
            let rem = PrayerNotificationContent.beforeAdhanReminderCopy(prayerKey: "maghrib", isFriday: false, minutes: 10)
            await enqueueTestNotification(
                title: rem.title,
                body: rem.body,
                categoryIdentifier: PrayerNotificationContent.CategoryID.reminder,
                sound: PrayerNotificationContent.sound(for: settings.notifications, channel: .reminder),
                userInfo: PrayerNotificationContent.debugUserInfo(kind: .reminderBeforeAdhan, prayerKey: "maghrib", mosqueSlug: slug)
            )
        }
    }

    /// Immediate delivery (`trigger: nil`), unique identifiers per fire — same approach as Al Muraja’ah / QuranScroll test hooks.
    private func enqueueTestNotification(
        title: String,
        body: String,
        categoryIdentifier: String,
        sound: UNNotificationSound,
        userInfo: [AnyHashable: Any]
    ) async {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = sound
        content.categoryIdentifier = categoryIdentifier
        content.userInfo = userInfo
        let identifier = "masjidly.debug.\(UUID().uuidString)"
        let request = UNNotificationRequest(identifier: identifier, content: content, trigger: nil)
        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
            UNUserNotificationCenter.current().add(request) { _ in
                continuation.resume()
            }
        }
    }

    private func applyNotificationPolicy() async {
        let n = settings.notifications
        if n.masterEnabled, let slug = settings.selectedMosqueSlug,
           let mosque = mosques.first(where: { $0.slug == slug }) ?? mosques.first(where: { $0.id == settings.selectedMosqueId }) {
            try? await notificationScheduler.rescheduleUpcomingPrayerNotifications(
                mosque: mosque,
                days: 7,
                settings: n,
                locale: settings.resolvedLocale
            )
        } else {
            await notificationScheduler.cancelAllPrayerNotifications()
        }
    }
}

File: apps/android/app/src/main/kotlin/com/mikhailspeaks/masjidly/features/settings/SettingsViewModel.kt
package com.mikhailspeaks.masjidly.features.settings

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mikhailspeaks.masjidly.data.SettingsStore
import com.mikhailspeaks.masjidly.data.cache.PrayerTimesDiskCache
import com.mikhailspeaks.masjidly.domain.MonthName
import com.mikhailspeaks.masjidly.domain.Mosque
import com.mikhailspeaks.masjidly.domain.PrayerRepository
import com.mikhailspeaks.masjidly.domain.PrayerTimesEngine
import com.mikhailspeaks.masjidly.features.notifications.PrayerNotificationContent
import com.mikhailspeaks.masjidly.features.notifications.PrayerNotificationPresenter
import com.mikhailspeaks.masjidly.features.notifications.PrayerNotificationScheduler
import kotlinx.coroutines.launch

/** Android counterpart to iOS `SettingsViewModel.swift`. */
class SettingsViewModel(
    private val repository: PrayerRepository,
    private val settings: SettingsStore,
    private val diskCache: PrayerTimesDiskCache,
    private val notificationScheduler: PrayerNotificationScheduler,
) : ViewModel() {
    var supportsMultipleAsrAdhan by mutableStateOf(false)
        private set

    var mosques: List<Mosque> = emptyList()
        private set

    init {
        refreshAsrAdhanSupport()
    }

    fun refreshAsrAdhanSupport() {
        viewModelScope.launch {
            val slug = settings.selectedMosqueSlug
                ?: settings.selectedMosqueId?.let { id ->
                    diskCache.loadMosques()?.firstOrNull { it.id == id }?.slug
                }
                ?: return@launch

            val parts = PrayerTimesEngine.getDateInSheffield(java.time.Instant.now())
            val month = MonthName.from(parts.month) ?: return@launch

            supportsMultipleAsrAdhan = try {
                val monthly = repository.getMonthlyPrayerTimes(slug, month, parts.year)
                if (monthly != null) {
                    diskCache.saveMonthly(slug, month.rawValue, parts.year, monthly)
                }
                val resolved = monthly ?: diskCache.loadMonthly(slug, month.rawValue, parts.year)
                resolved?.prayerTimes?.any { !it.asrMithl2.isNullOrEmpty() } == true
            } catch (_: Exception) {
                val cached = diskCache.loadMonthly(slug, month.rawValue, parts.year)
                cached?.prayerTimes?.any { !it.asrMithl2.isNullOrEmpty() } == true
            }
        }
    }

    fun updateMosques(list: List<Mosque>) {
        mosques = list
    }

    fun onNotificationsChanged() {
        viewModelScope.launch { applyNotificationPolicy() }
    }

    fun fireTestNotification(context: Context, type: TestNotificationType) {
        viewModelScope.launch {
            if (!notificationScheduler.requestAuthorizationIfNeeded()) return@launch
            val language = settings.appLanguage
            val slug = settings.selectedMosqueSlug.orEmpty()
            when (type) {
                TestNotificationType.ADHAN -> {
                    val copy = PrayerNotificationContent.adhanCopy("maghrib", isFriday = false, language)
                    PrayerNotificationPresenter.showInstantTest(
                        context = context,
                        title = copy.first,
                        body = copy.second,
                        categoryId = PrayerNotificationContent.CategoryId.ADHAN,
                        extras = PrayerNotificationContent.debugUserInfo(
                            PrayerNotificationContent.PayloadKind.ADHAN,
                            "maghrib",
                            slug,
                        ),
                    )
                }
                TestNotificationType.IQAMAH -> {
                    val copy = PrayerNotificationContent.iqamahCopy("maghrib", isFriday = false, language)
                    PrayerNotificationPresenter.showInstantTest(
                        context = context,
                        title = copy.first,
                        body = copy.second,
                        categoryId = PrayerNotificationContent.CategoryId.IQAMAH,
                        extras = PrayerNotificationContent.debugUserInfo(
                            PrayerNotificationContent.PayloadKind.IQAMAH,
                            "maghrib",
                            slug,
                        ),
                    )
                }
                TestNotificationType.REMINDER -> {
                    val copy = PrayerNotificationContent.beforeAdhanReminderCopy(
                        "maghrib",
                        isFriday = false,
                        minutes = 10,
                        language,
                    )
                    PrayerNotificationPresenter.showInstantTest(
                        context = context,
                        title = copy.first,
                        body = copy.second,
                        categoryId = PrayerNotificationContent.CategoryId.REMINDER,
                        extras = PrayerNotificationContent.debugUserInfo(
                            PrayerNotificationContent.PayloadKind.REMINDER_BEFORE_ADHAN,
                            "maghrib",
                            slug,
                        ),
                    )
                }
                TestNotificationType.ALL -> {
                    fireTestNotification(context, TestNotificationType.ADHAN)
                    fireTestNotification(context, TestNotificationType.IQAMAH)
                    fireTestNotification(context, TestNotificationType.REMINDER)
                }
            }
        }
    }

    private suspend fun applyNotificationPolicy() {
        val n = settings.notifications
        val mosque = settings.selectedMosqueSlug?.let { slug ->
            mosques.firstOrNull { it.slug == slug }
        } ?: settings.selectedMosqueId?.let { id ->
            mosques.firstOrNull { it.id == id }
        }
        if (n.masterEnabled && mosque != null) {
            notificationScheduler.rescheduleUpcomingPrayerNotifications(
                mosque = mosque,
                days = 7,
                settings = n,
                language = settings.appLanguage,
                asrIqamahPreference = settings.asrIqamahPreference,
            )
        } else {
            notificationScheduler.cancelAllPrayerNotifications()
        }
    }

    enum class TestNotificationType {
        ADHAN,
        IQAMAH,
        REMINDER,
        ALL,
    }
}
